Thread analyzer adsp fix

jsarha opened the pull request:
Due to Intel ADSP's cache architecture the Zephyr thread analyzer does not work out of the box on SOF platform. The core specific caches are not synchronized between the cores and accessing thread date from other core usually causes a crash. This PR adds on option for thread analyzer thread to only "analyze" threads on the same core its running on, and starts thread for each core.

Commit:
kernel: thread: Add k_thread_foreach_filter_by_cpu() and unlocked version

Add functions k_thread_foreach_unlocked_filter_by_cpu() and
k_thread_foreach_filter_by_cpu() to loop through the threads on the
specified cpu only.

Signed-off-by: Jyri Sarha <jyri.sarha@linux.intel.com>

Commit:
kernel: thread: Add k_thread_runtime_stats_cpu_get()

Add k_thread_runtime_stats_cpu_get() to get runtime statistics of
the specified core.

Signed-off-by: Jyri Sarha <jyri.sarha@linux.intel.com>

Commit:
debug: thread_analyzer: Option to analyze threads on each core separately

Add implementation to analyze threads on each cpu separately. This
feature can be enabled with THREAD_ANALYZER_AUTO_SEPARATE_CORES Kconfig
option. If enabled, an analyzer thread is started for each cpu, and
the threads will only analyze thread on the cpu its running on.

This feature is needed for Intel ADSP platform, where cpu specific
caches are not synchronized between the cpu. It is also probably
needed by other platforms having CONFIG_KERNEL_COHERENCE=y, so default
to THREAD_ANALYZER_AUTO_SEPARATE_CORES=y for those platform.

Signed-off-by: Jyri Sarha <jyri.sarha@linux.intel.com>

kv2019i commented:
I think the code look good for wider review. I think the rationale could be expanded a bit. Which piece of data was problematic without this patch? Zephyr allocates most of the its data uncached on xtensa as well, so a bit surprising the caches are causing a problem here on Zephyr side. E.g. global static data is uncached by default and should be safe to use across cores (evein on these xtensa platforms). It would help to pinpoint the exact data object (and where it is accessed) that is causing problems, and explain why these additions are needed.

jsarha commented:
> I think the code look good for wider review. I think the rationale could be expanded a bit. Which piece of data was problematic without this patch? Zephyr allocates most of the its data uncached on xtensa as well, so a bit surprising the caches are causing a problem here on Zephyr side. E.g. global static data is uncached by default and should be safe to use across cores (evein on these xtensa platforms). It would help to pinpoint the exact data object (and where it is accessed) that is causing problems, and explain why these additions are needed.

The kernel thread list can be looped through, but it appeared that calling k_thread_runtime_stats_all_get() hung the DSP (I never got any crash, the DSP just became dead), also asking isr stack from neighboring core also hung the DSP.

The was a build error for non SMP build, I'll try to fix all the automatic testing issues and the mark this ready for review.

andyross reviewed (commented):
@kv2019i Surely the problem here is that on SMP builds with adsp, the thread stacks which this utility needs to inspect are in cached/CPU-local memory.  So the data it sees will be an essentially arbitrary mix of stale, previously-flushed cache blocks and not the sequentially consistent data it expects.  But agree that the comments should be clearer on what this is doing.

Nitpick-wise: Might be worth tying this directly to CONFIG_KERNEL_COHERENCE, which is the actual feature that enables the cache-incoherent thread stacks.  If you don't have that =y there's no need for the per-CPU thread analyzer.

Likewise this is using CONFIG_SCHED_CPU_MASK APIs, which is an optional feature that should be listed in a "depends on" in kconfig.

But on the whole this is a reasonable and straightforward workaround for the problem.

jsarha commented on subsys/debug/thread_analyzer.c:
```diff
@@ -129,61 +136,127 @@ static void thread_analyze_cb(const struct k_thread *cthread, void *user_data)
+
+static K_THREAD_STACK_ARRAY_DEFINE(analyzer_thread_stacks, CONFIG_MP_MAX_NUM_CPUS,
+				   CONFIG_THREAD_ANALYZER_AUTO_STACK_SIZE);
+static struct k_thread analyzer_thread[CONFIG_MP_MAX_NUM_CPUS];
```
It should in theory be possible to survive with two threads. One all the time at core 0, and the other being pinned by the core 0 thread different cores, synchronized with semaphores, and having some timeout mechanism to skip suspended cores... but then again I don't think the save is worth all that complexity.

andyross replied:
FWIW, the savings is very significant.  A thread stack is the largest kernel data structure you're going to allocate.   The DSPs are big but ~8k of stack memory on MTL is still a big waste.  And again, I don't see that this is so complicated?  Just pend the thread for each change of CPU and bounce to a k_timer to do the migration.  You'll pay a cost of 0-1 ticks (MTL is still 12kHz ticks I think?), which seems reasonable for a debug feature.

jsarha commented:
> @kv2019i Surely the problem here is that on SMP builds with adsp, the thread stacks which this utility needs to inspect are in cached/CPU-local memory. So the data it sees will be an essentially arbitrary mix of stale, previously-flushed cache blocks and not the sequentially consistent data it expects. But agree that the comments should be clearer on what this is doing.
> 
> Nitpick-wise: Might be worth tying this directly to CONFIG_KERNEL_COHERENCE, which is the actual feature that enables the cache-incoherent thread stacks. If you don't have that =y there's no need for the per-CPU thread analyzer.
> 
I tried setting CONFIG_KERNEL_COHERENCE=n and taking away my changes, but the FW still crashed with the change. So there is something else to it than just that. 

> Likewise this is using CONFIG_SCHED_CPU_MASK APIs, which is an optional feature that should be listed in a "depends on" in kconfig.
> 

Thanks, I'll add the dependency.

> But on the whole this is a reasonable and straightforward workaround for the problem.

jsarha commented:
@nashif , Liam suggested me to ping you about this PR.

peter-mitsis commented on include/zephyr/kernel.h:
```diff
@@ -122,6 +122,33 @@ typedef void (*k_thread_user_cb_t)(const struct k_thread *thread,
+ * threads are blocked until this API returns.
+ */
+#ifdef CONFIG_SMP
+void k_thread_foreach_my_core(k_thread_user_cb_t user_cb, void *user_data);
```
Instead of using 'xxx_my_core', it might be more consistent to use 'xxx_current_cpu'.

peter-mitsis commented on kernel/thread_monitor.c:
```diff
@@ -83,3 +83,56 @@ void k_thread_foreach_unlocked(k_thread_user_cb_t user_cb, void *user_data)
+	SYS_PORT_TRACING_FUNC_ENTER(k_thread, foreach);
+
+	for (thread = _kernel.threads; thread; thread = thread->next_thread) {
+		if (thread->base.cpu == arch_proc_id())
```
I have three issues with `if (thread->base.cpu == arch_proc_id())`.

First -- arch_proc_id(). Although this returns a unique identifier per CPU, it is not guaranteed to match the cpu ID numbering that is used within the zephyr kernel. As a result, when thread->base.cpu is set in kernel/include/kswap.h::do_swap() via `new_thread->base.cpu = arch_curr_cpu()->id` the IDs may match on some platforms, but not all.

Second --  Near as I can tell, thread->base.cpu is only set in do_swap() which is called when a thread does a context switch. Unless I have missed it, it does not look like it gets updated when exiting an interrupt. Thus, base.cpu is not currently reliable.

Third -- CPU Masking. base.cpu identifies the last CPU on which the thread executed. If CPU masking is enabled and the set of CPU to which the thread is pinned changed while it was blocked on some kernel object and it has not yet had a chance to run on the new CPU, does it really belong with the threads associated with its old CPU?

peter-mitsis replied:
FYI, the second issue I identified above has been fixed and merged.

jsarha replied:
As you know the Intel ADSP is quite unusual Zephyr platform, and this change quite possibly is only needed for it. So I wonder if we should make the new "current_cpu"-functions optional on some new Kconfig option, and possibly make that option dependent on SOC_FAMILY_INTEL_ADSP?

Then the practical side. I so what should I use here, if not (thread->base.cpu == arch_proc_id())? Would calculating the arch_curr_cpu() offset in struct z_kernel -> cpus[] table be better?

andyross commented:
> I tried setting CONFIG_KERNEL_COHERENCE=n and taking away my changes, but the FW still crashed with the change.

No, I meant predicate this feature so that it gets enabled when COHERENCE=y.  You can't meaningfully run SMP kernels on intel_adsp without COHERENCE (well, you can, but it maps all memory as uncached -- this used to work but might have bitrotten).

jsarha commented:
Changes since previous version:
- `arch_curr_cpu()` to `_current_cpu->id`
- *my_core* to *current_cpu*
- add "default y if KERNEL_COHERENCE" to THREAD_ANALYZER_AUTO_SEPARATE_CORES
- some more text to commit messages and Kconfig help text

andyross reviewed (approved):
I'm out of complaints.  This looks correct to me.  I'm still mildly annoyed by all the extra threads, but this is after all a debug feature and not a production thing.  If it fits, it fits, I guess.

jsarha commented:
Oops, using `_current_cpu->id` was not such a good idea, can't access _current_cpu from user-space side.  Use `arch_curr_cpu()->id` instead. Part of the reason for pushing these changes is to implement Thread Analyzer kind of functionality outside Zephyr as a SOF feature. If anyone is interested, one can take a look here: https://github.com/thesofproject/sof/pull/9084

andyross commented:
> Oops, using _current_cpu->id was not such a good idea, can't access _current_cpu from user-space side. Use arch_curr_cpu()->id instead

You can't do that from a userspace thread either.  Both are accessing the CPU struct, which is in kernel memory.   The solution if you want to call this from userspace is to make it a system call.

FWIW: behaviorally those are exactly the same code.  The difference is that _current_cpu has an assertion in it to check that your context is not "mobile" (i.e. that you won't end up on another CPU before you use the result of the call).  You must be in an ISR or have interrupts masked on the local CPU (which generally means "in a spinlock").  I was pretty sure when I reviewed that you were doing it correctly, no?

andyross reviewed (changes requested):
Flip vote: no seriously, you can't be evading the z_smp_cpu_mobile() requirement without introducing race conditions vs. CPU migration.

npitre commented:
>> Oops, using _current_cpu->id was not such a good idea, can't access
>> _current_cpu from user-space side. Use arch_curr_cpu()->id instead
>
> You can't do that from a userspace thread either.  Both are accessing the CPU
> struct, which is in kernel memory.   The solution if you want to call this
> from userspace is to make it a system call.

And what would be the point? By the time the system call returns, you might
already be running on another CPU than the one reported.

jsarha commented:
> And what would be the point? By the time the system call returns, you might already be running on another CPU than the one reported.

Not on a system where this feature is needed, at least not on Intel ADSP. There, in order to get benefit from uncoherent per cpu cache, all processes are pinned to single cpu only. That also the reason why thread analyzer thread can not go around peeking the thread statistics from a neighboring cpu, but each cpu should have their own thread analyzer thread.

andyross commented:
@npitre Well, at least in principle an app might be using cpu_mask (e.g. SOF is built with SCHED_CPU_MASK_PIN_ONLY) to prevent migration, and would be able to get reliable output using this API.  In point of fact the whole thread_analyzer feature sort of accepts that it might need to be used in a race-prone manner, because doing this reliably requires holding a global spinlock around the thread list, which is an extremely high cost to pay.  It's a debug feature after all.

npitre commented:
> There, in order to get benefit from uncoherent per cpu cache, all 
> processes are pinned to single cpu only. That also the reason why thread
> analyzer thread can not go around peeking the thread statistics from a
> neighboring cpu, but each cpu should have their own thread analyzer thread.

Then why don't you simply pass each pinned thread its CPU number as an extra
argument? Problem solved, no deceptive syscall needed.

jsarha commented:
> Then why don't you simply pass each pinned thread its CPU number as an extra
> argument? Problem solved, no deceptive syscall needed.

Sure, I can do that. Then, instead of adding `k_thread_foreach_current_cpu()` I would need to implement something like `k_thread_foreach_filter_by_cpu(k_thread_user_cb_t user_cb, void *user_data, int cpu)`. Would you like that better?

jsarha commented:
> > Then why don't you simply pass each pinned thread its CPU number as an extra
> > argument? Problem solved, no deceptive syscall needed.
> 
> Sure, I can do that. Then, instead of adding `k_thread_foreach_current_cpu()` I would need to implement something like `k_thread_foreach_filter_by_cpu(k_thread_user_cb_t user_cb, void *user_data, int cpu)`. Would you like that better?
> 
> Oh, and I would need `k_thread_runtime_stats_get_by_cpu()` or something too.

Applied the change suggested here.

jsarha commented:
> > There, in order to get benefit from uncoherent per cpu cache, all
> > processes are pinned to single cpu only. That also the reason why thread
> > analyzer thread can not go around peeking the thread statistics from a
> > neighboring cpu, but each cpu should have their own thread analyzer thread.
> 
> Then why don't you simply pass each pinned thread its CPU number as an extra argument? Problem solved, no deceptive syscall needed.

@npitre this is now done. Is there something more I should do here to get this merged?

jsarha commented:
Oops, k_thread_foreach_filter_by_cpu() implementation had wrong name. Apparently I only tested the unlocked version. Now fixed.

kv2019i commented on kernel/thread_monitor.c:
```diff
@@ -83,3 +83,60 @@ void k_thread_foreach_unlocked(k_thread_user_cb_t user_cb, void *user_data)
+	SYS_PORT_TRACING_FUNC_ENTER(k_thread, foreach);
+
+	for (thread = _kernel.threads; thread; thread = thread->next_thread) {
+		if (thread->base.cpu == arch_curr_cpu()->id)
```
"cpu" is passed as argument, but filtering is done by arch_curr_cpu() ...? Either the argument is missing or implementation is not correct.

jsarha replied:
Uh, thanks. I must have simply forgot to change that.

kv2019i commented on kernel/thread_monitor.c:
```diff
@@ -83,3 +83,60 @@ void k_thread_foreach_unlocked(k_thread_user_cb_t user_cb, void *user_data)
+	SYS_PORT_TRACING_FUNC_ENTER(k_thread, foreach_unlocked);
+
+	for (thread = _kernel.threads; thread; thread = thread->next_thread) {
+		if (thread->base.cpu == cpu) {
```
Here filtering is done based on "cpu" passed as argument.

kv2019i commented on subsys/debug/thread_analyzer.c:
```diff
@@ -129,61 +143,132 @@ static void thread_analyze_cb(const struct k_thread *cthread, void *user_data)
+			k_thread_abort(tid);
+			continue;
+		}
+		snprintf(name, sizeof(name), "core %u thread anlyzer", i);
```
typo: anlyzer

kv2019i reviewed (commented):
@jsarha Latter two patches ok, but one clarifying question on first patch. There seems to be a semantics difference between locked and unlocked versions. Am I missing something?

kv2019i reviewed (approved):
Thanks, looks good now!

jsarha commented:
@andyross , there was one-line bug-fix, but otherwise this should be the same as the version you approved.

jsarha commented:
Argh, forgot about that this is still lying around. @andyross , a gentle ping!

 @lgirdwood I'am not sure how to get this moving.

jsarha commented:
@peter-mitsis , could you spare time to review this?

peter-mitsis commented:
@jsarha - Taking a look at this again.

andyross reviewed (approved):
Still looks good here

aescolar commented:
@jsarha @andyross @nashif is this meant to be backported to 3.7?

jsarha commented:
> @jsarha @andyross @nashif is this meant to be backported to 3.7?

@aescolar No, we should not have any need back-porting, but I would very much like to see this merged for the next release.

andyross commented:
> @jsarha @andyross @nashif is this meant to be backported to 3.7?

@kv2019i @lyakh might have input too.  But broadly SOF tracks Zephy upstream closely.  At least so far there hasn't been an attempt to tie a particular SOF release to a Zephyr LTS.  And honestly even if there were, this is a fix to a debug feature that isn't used in production audio firmware, so seems like it would be low priority.

(Just FWIW: I'm fairly conservative about bugfix backports.  Security bugs, 100%.  Bugs reported by actual shipping apps, absolutely.  But just general fixes?  There are too many and IMHO the churn caused by the fixes themselves are a higher risk than the bugs they fix in most cases.)
